hbase-server/src/main/java/org/apache/hadoop/hbase/master/TableStateManager.java

```diff
@@ -54,8 +53,20 @@
 // TODO: Make this a guava Service
 @InterfaceAudience.Private
 public class TableStateManager {
-
   private static final Logger LOG = LoggerFactory.getLogger(TableStateManager.class);
+
+  /**
+   * All table state is kept in hbase:meta except that of hbase:meta itself.
```

Comment:
In general I do not think this is a good design. We should persist it somewhere, and just cache it in master's memory. And we do not need to disable a table when altering any more? And the solution here just assume that we only have one meta region? So the altering operation can be atomic? What if we have multiple meta regions and we crash in the middle? I think this patch also aims to implement splittable meta in the future? No?

Reply:
bq. We should persist it somewhere, and just cache it in master's memory.

I explain why state is intentionally transient for now so table falls-back to ENABLED if problem like Master crash during alter; its one less thing for the operator to decipher and hbck2 navigate starting up a cluster; in other words we fallback to the 'known' state rather than persist DISABLED meta, a new condition that is sure to have holes in it.

bq. And the solution here just assume that we only have one meta region? 

All of the code base presumes this, not just this patch (this patch is part work to undo this presumption letting go of hardcoded meta schema).

bq. I think this patch also aims to implement splittable meta in the future? No?

Splittable meta is another project, not this one.

bq. And we do not need to disable a table when altering any more? 

This patch makes hbase:meta dynamic. Maybe the above will work. Let me try.

Reply:
I know splittable meta is not implemented by this one but I think we should prepare for it? Not adding new difficulty, by introducing more 'there is only one meta region' assumptions...

And I do not think we need to disable a table before altering it. The issue here is named 'Altering meta', not 'disabling meta', so let's try remove the disable/enable stuff and just implement the alter?

Reply:
This discussion has been by-passed by your suggestion that we just support alter of meta, and not disable.